### PR TITLE
[frameit] Take in account effective_text_height for available text height

### LINKED ISSUE #14625: [Frameit] Multiline text is unnecessarily shrunk
🔑

### New Issue Checklist

- [x] Updated fastlane to the latest version
- [ ] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When framing screenshots with multilines, text is shrunk unexpectedly while using the title_min_height parameter.

#### Here is the result of a framing a one-line text:
iPhone X-01_scan_framed copy

#### Here is the result of framing a two-line text:

iPhone X-03_multipage_pdf_framed copy

There was no need to shrink the text: even without shrinking the text the padding was respected.

#### Expected result:
iPhone X-06_export_framed copy

#### Why?
The issue comes from this line: https://github.com/fastlane/fastlane/blob/master/frameit/lib/frameit/editor.rb#L341. It incorrectly assumes that the height available for the text is necessarily one font size. It should take in account the title_min_height parameter. Even if not taking in account the title_min_height, I'm not sure it should assume that the height available for the text is one font size, because that doesn't work for multiline.

Framefile.json:

```
{
  "device_frame_version": "latest",
  "default": {
    "title": {
      "color": "#ffffff",
      "fonts" : [
      {
          "font": "../fonts/zh-Hans.otf",
          "supported": ["zh-Hans"]
        },
        {
          "font": "../fonts/ja.otf",
          "supported": ["ja"]
        },
        {
          "font": "../fonts/zh-Hant.otf",
          "supported": ["zh-Hant"]
        }
      ]
    },
    "background": "./Frameit-background.png",
    "padding": 40,
    "show_complete_frame": false,
    "stack_title" : true,
    "title_below_image": false,
    "font_scale_factor": 0.05,
    "title_min_height" : "20%"
  },
   "data": [
    {
      "filter": "scan"
    },
    {
      "filter": "edit"
    },
    {
      "filter": "multipage_pdf"
    },
    {
      "filter": "organize"
    },
    {
      "filter": "export"
    },
    {
      "filter" : "document"
    },
    {
      "filter" : "iPad Pro",
      "show_complete_frame" : true,
      "font_scale_factor": 0.035
    },
    {
      "filter" : "iPhone X",
      "title_min_height" : "30%"
    }
  ]
}
```

#### Fix
I will try to make the corresponding PR. 

It seems to me that the frameit editor engine code is quite fragile and does too much, making it hard to maintain or update, so I will make a very minimal change.

A suggestion would be to separate the layout engine from the actual image editing. We should be able to do a first pass where we compute the different element locations, and apply it in a second pass.

#### Documentation
We should probably document the title_min_height parameter as well. I think that was my bad.

### PR DESCRIPTION
**Pull request by bvirlet**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Fix #14625

### CODE REVIEW

File: frameit/lib/frameit/editor.rb [hunk 1]
```diff
@@ -338,7 +338,7 @@ def put_title_into_background(background, stack_title)
       # Resize the 2 labels if they exceed the available space either horizontally or vertically:
       image_scale_factor = 1.0 # default
       ratio_horizontal = sum_width / (image.width.to_f - horizontal_frame_padding * 2) # The fraction of the text images compared to the left and right padding
-      ratio_vertical = title.height.to_f / actual_font_size # The fraction of the actual height of the images compared to the available space
+      ratio_vertical = title.height.to_f / (title_min_height || actual_font_size) # The fraction of the actual height of the images compared to the available space
```
Comment (joshdholtz) [inline frameit/lib/frameit/editor.rb]: Should we be using `effective_text_height` instead of `title_min_height`? 🤔

Comment (bvirlet) [reply frameit/lib/frameit/editor.rb]: OK. I had initially done that thinking that `effective_text_height` was not computed yet but this assumption was incorrect.

**Comment by fbeeper** [comment]
I can't effectively review this PR (neither I'm a reviewer, nor I'm well versed in ruby or frameit code), but let me add some notes having tried other fixes, and having tested this one change locally:
- Works great as a fix for #14625.
- Respects vertical padding as expected.
- Doesn't seem to change single-line text rendering.

**Comment by janpio** [comment]
That's awesome @fbeeper, thanks.

Do you by chance have a minimal frameit configuration we could use to test the before/after result with and without this PR? That would greatly help us "real" reviewers (not sure if blessing or a curse 👼 👿 ;) ) to get this approved.

PS: You can still leave you positive review as "Approve" in the "Files changed" tab here on Github, it just won't be highlighted green but grey.

**Comment by bvirlet** [comment]
Thanks @fbeeper for looking into it.

**Comment by bvirlet** [comment]
@janpio Here is an example project that reproduces the issue and the fix:
[example.zip](https://github.com/fastlane/fastlane/files/3211841/example.zip)

File: frameit/lib/frameit/editor.rb [hunk 2]
```diff
@@ -338,7 +338,7 @@ def put_title_into_background(background, stack_title)
       # Resize the 2 labels if they exceed the available space either horizontally or vertically:
       image_scale_factor = 1.0 # default
       ratio_horizontal = sum_width / (image.width.to_f - horizontal_frame_padding * 2) # The fraction of the text images compared to the left and right padding
-      ratio_vertical = title.height.to_f / actual_font_size # The fraction of the actual height of the images compared to the available space
+      ratio_vertical = title.height.to_f / (effective_text_height || actual_font_size) # The fraction of the actual height of the images compared to the available space
```
Comment (joshdholtz) [inline frameit/lib/frameit/editor.rb]: After looking at the code a bit, it looks like `effective_text_height` will probably never be `nil` so do we just make it look like this??? 👇
```suggestion
      ratio_vertical = title.height.to_f / (effective_text_height) # The fraction of the actual height of the images compared to the available space
```

**Review by joshdholtz — COMMENTED** [review]
Works 💯 but wondering if we can remove reference to `actual_font_size` since it doesn't look like it will ever be used 🙃

**Comment by bvirlet** [comment]
@joshdholtz looks like the tests failed on MacOS for a unknown reason. Probably just need to restart the build but I cannot do that myself.

**Comment by janpio** [comment]
Restart won't be enough here, Circle CI deprecated the Xcode image we were using. Rebasing onto `master` should fix the problem though.

**Comment by bvirlet** [comment]
CI build fixed!

**Comment by bvirlet** [comment]
Hey @janpio, do you think this could be merged?

**Comment by janpio** [comment]
I only merge things that I really tested and approved, and I don't have the capacity to do that right now unfortunately (life...) - sorry.

Let's hope @joshdholtz can take another look soon, I think he tested it before and just didn't follow up after you made the changes.

**Comment by bvirlet** [comment]
@janpio oops, sorry, I meant @joshdholtz.

No worries about not having the time, and no need to be sorry, life first !!! I'm already so grateful of the time you spend helping with fastlane. 🤗

**Comment by bvirlet** [comment]
@joshdholtz politely requesting a final review and merging whenever you find the time. I have addressed your previous review comments and the tests are passing.

Have a great week-end!

**Comment by bvirlet** [comment]
Hey @joshdholtz ! Just wondering if you would have the time for a review just to make sure this PR doesn't get lost as the feature could be quite useful !

**Review by fbeeper — APPROVED** [review]

**Comment by bvirlet** [comment]
Hi @joshdholtz, this is still quite relevant for us and other folks. Is there anything I could do to help to get this merged? Thanks!

**Comment by joshdholtz** [comment]
Ahhh, my bad! I thought we we merged this 😱 I will take another look after I get my first cup of coffee

**Comment by joshdholtz** [comment]
@bvirlet I got my coffee finally!

**Review by joshdholtz — APPROVED** [review]
Looks good to me! :shipit:

**Pull request merged**